pkg/runtime/grpc/grpc-runtime.go

```diff
@@ -0,0 +1,440 @@
+		// return NewK8SPortForwardConn(ctx, s, time.Second*30)
+	})
+
+	conn, err := grpc.DialContext(ctx, "", dialOpt, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+	if err != nil {
+		return nil, fmt.Errorf("could not dial gadget pod on node %q: %w", pod.node, err)
+	}
+	defer conn.Close()
+	client := pb.NewGadgetManagerClient(conn)
+
+	allParams := make(map[string]string)
+	gadgetCtx.GadgetParams().CopyToMap(allParams, "")
+	gadgetCtx.OperatorsParamCollection().CopyToMap(allParams, "operator.")
+
+	runRequest := &pb.GadgetRunRequest{
+		GadgetName:     gadgetCtx.GadgetDesc().Name(),
+		GadgetCategory: gadgetCtx.GadgetDesc().Category(),
+		Params:         allParams,
+		Nodes:          nil,
+		FanOut:         false,
+		LogLevel:       uint32(gadgetCtx.Logger().GetLevel()),
+	}
+
+	runClient, err := client.RunGadget(ctx)
+	if err != nil && !errors.Is(err, context.Canceled) {
+		return nil, err
+	}
+
+	controlRequest := &pb.GadgetControlRequest{Event: &pb.GadgetControlRequest_RunRequest{RunRequest: runRequest}}
+	err = runClient.Send(controlRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	parser := gadgetCtx.Parser()
+
+	jsonHandler := func([]byte) {}
+	jsonArrayHandler := func([]byte) {}
+
+	if parser != nil {
+		jsonHandler = parser.JSONHandlerFunc()
+		jsonArrayHandler = parser.JSONHandlerFuncArray(pod.node)
+	}
+
+	doneChan := make(chan bool)
+
+	var result []byte
+	expectedSeq := uint32(1)
+
+	go func() {
+		for {
+			ev, err := runClient.Recv()
+			if err != nil {
+				break
+			}
+			switch ev.Type {
+			case pb.EventTypeGadgetPayload:
+				if len(ev.Payload) > 0 && ev.Payload[0] == '[' {
+					jsonArrayHandler(ev.Payload)
+					continue
+				}
+				if expectedSeq != ev.Seq {
+					gadgetCtx.Logger().Warnf("expected seq %d, %d messages dropped", expectedSeq, ev.Seq-expectedSeq)
+				}
+				expectedSeq = ev.Seq + 1
+				jsonHandler(ev.Payload)
+			case pb.EventTypeGadgetResult:
+				result = ev.Payload
+			case pb.EventTypeGadgetJobID:
+			// not needed right now
+			case pb.EventTypeGadgetDone:
+				gadgetCtx.Logger().Debug("got EventTypeGadgetDone from server")
+				doneChan <- true
+				return
+			default:
+				if ev.Type >= 1<<pb.EventLogShift {
+					gadgetCtx.Logger().Log(logger.Level(ev.Type>>pb.EventLogShift), fmt.Sprintf("%-20s | %s", pod.node, string(ev.Payload)))
+					continue
+				}
+				gadgetCtx.Logger().Warnf("unknown payload type %d: %s", ev.Type, ev.Payload)
+			}
+		}
+		doneChan <- true
+	}()
+
+	select {
+	case <-doneChan:
+		gadgetCtx.Logger().Debug("done from server side")
+	case <-gadgetCtx.Context().Done():
+		// Send stop request
+		controlRequest := &pb.GadgetControlRequest{Event: &pb.GadgetControlRequest_StopRequest{StopRequest: &pb.GadgetStopRequest{}}}
+		runClient.Send(controlRequest)
+
+		// Wait for done or timeout
+		select {
+		case <-doneChan:
+			gadgetCtx.Logger().Debug("done after cancel request")
+		case <-time.NewTimer(ResultTimeout * time.Second).C:
```

Comment:
Because that's the channel it waits for. But I changed it to `time.After()` now, which is a shortcut to the same behavior and probably what you meant as well.